Fixed endianness on IPv4 and iPv6 struct packing

apognu opened the pull request:
Relates to #15.

This changes how you pack the structs for v4 and v6 addresses. Since some fields must be represented as BigEndian and some other as LittleEndian. I'm not sure if this is the proper way or if there was a way to keep your unsafe pointer method and still get to the same output.

You tell me.

mdlayher commented on internal/wgnl/configure_linux.go:
```diff
@@ -210,26 +211,27 @@ func sockaddrBytes(endpoint net.UDPAddr) ([]byte, error) {
-			Addr:   addr,
-		}
+		buf := bytes.NewBuffer(nil)
+		binary.Write(buf, binary.LittleEndian, uint16(unix.AF_INET6))
```
This won't be correct on big endian architectures, right?  You can do `ne := nlenc.NativeEndian()` and pass that for the fields that are endian-specific.

mdlayher commented on internal/wgnl/configure_linux.go:
```diff
@@ -210,26 +211,27 @@ func sockaddrBytes(endpoint net.UDPAddr) ([]byte, error) {
-			Port:   uint16(endpoint.Port),
-			Addr:   addr,
-		}
+		buf := bytes.NewBuffer(nil)
```
Leave a comment that this structure layout matches unix.RawSockaddrInet6.

mdlayher commented on internal/wgnl/configure_linux.go:
```diff
@@ -210,26 +211,27 @@ func sockaddrBytes(endpoint net.UDPAddr) ([]byte, error) {
-		Port:   uint16(endpoint.Port),
-		Addr:   addr,
-	}
+	buf := bytes.NewBuffer(nil)
```
Leave a comment that this structure layout matches unix.RawSockaddrInet4.

mdlayher commented:
I assume a similar fix has to be applied on the parsing side?

apognu commented:
I am currently reading the tests and trying to make sense of them.

I will add the requested comments.

apognu commented:
So, I reverted my changes to your original method (using plain unix.* structs), and just converted the endpoint port to big-endian when necessary. Do you think that's better?

mdlayher commented:
That's a clever idea; I imagine it'll work just fine.  The port is "wrong" in the userspace perspective but the kernel will interpret the value as big endian and it ought to just work!

Thanks for your help so far.

apognu commented:
Parsing is fine in my last commit, and all tests should pass in configure and parse.

mdlayher commented on internal/wgnl/configure_linux.go:
```diff
@@ -212,7 +213,7 @@ func sockaddrBytes(endpoint net.UDPAddr) ([]byte, error) {
 		sa := unix.RawSockaddrInet6{
 			Family: unix.AF_INET6,
-			Port:   uint16(endpoint.Port),
+			Port:   binary.BigEndian.Uint16(nlenc.Uint16Bytes(uint16(endpoint.Port))),
```
Let's split this pattern out into a helper function:

```go
// sockaddrPort interprets port as a big endian uint16 for use passing sockaddr
// structures to the kernel.
func sockaddrPort(port int) uint16 {
    return binary.BigEndian.Uint16(nlenc.Uint16Bytes(uint16(port)))
}
```

Can add it near the bottom of this file.

mdlayher commented on internal/wgnl/parse_linux.go:
```diff
@@ -230,9 +232,11 @@ func parseSockaddr(endpoint *net.UDPAddr) func(b []byte) error {
 			// IPv4 address parsing.
 			sa := *(*unix.RawSockaddrInet4)(unsafe.Pointer(&b[0]))
 
+			fmt.Println()
```
Remove please!

mdlayher commented on internal/wgnl/configure_linux_test.go:
```diff
@@ -146,7 +147,7 @@ func TestLinuxClientConfigureDevice(t *testing.T) {
 											0x00, 0x00, 0x00, 0x33,
 										},
-										Port: 51820,
+										Port: binary.BigEndian.Uint16(nlenc.Uint16Bytes(51820)),
```
Sorry; can you use the helper in the tests too?  They're in the same package so should be fine.

mdlayher commented on internal/wgnl/parse_linux_test.go:
```diff
@@ -209,7 +210,7 @@ func TestLinuxClientDevicesOK(t *testing.T) {
 										Data: (*(*[unix.SizeofSockaddrInet4]byte)(unsafe.Pointer(&unix.RawSockaddrInet4{
 											Addr: [4]byte{192, 168, 1, 1},
-											Port: 1111,
+											Port: binary.BigEndian.Uint16(nlenc.Uint16Bytes(1111)),
```
Same.

apognu commented:
No problem, that's what it's for.

mdlayher commented:
Thanks again!
